Multiple aspects running py_binary based tools errors in Bazel 8

### Description of the bug:


Upon upgrading to Bazel 8.0.0, I started encountering an issue where two (or more) aspects that run a rules_python 1.0.0 py_binary as the executable are missing files. I see errors in the Python launcher script where the runfiles symlinks ultimately resolve to paths that do not, in fact, exist. The runfiles symlinks appear in the correct locations in the runfiles tree and are complete, but the locations the links point to are missing all but one of the aspect's underlying files.

<details>
<summary>More details/error messages:</summary>

When running a build I see errors like this:
```
ERROR: /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_pydeps++reqs+reqs/BUILD.bazel:7:15: Action external/rules_pydeps++reqs+reqs/pip_deps_index failed: (Exit 1): index failed: error executing Action command (from target @@rules_pydeps++reqs+reqs//:pip_deps_index) bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index ... (remaining 1 argument skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/2/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index", line 580, in <module>
    Main()
  File "/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/2/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index", line 504, in Main
    assert os.path.exists(main_filename), \
AssertionError: Cannot exec() '/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/2/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index.runfiles/_main/../rules_pydeps+/pydeps/private/index/index.py': file not found.
```

If I run with --sandbox_debug, and I inspect the sandbox, what I find is that the runfiles tree for index.runfiles looks like it should have the files:

```
❯ tree /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/6/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index.runfiles/_main/../rules_pydeps+/pydeps/
/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/6/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index.runfiles/_main/../rules_pydeps+/pydeps/
├── __init__.py
└── private
    ├── __init__.py
    ├── index
    │   ├── __init__.py -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/external/rules_pydeps+/pydeps/private/index/__init__.py
    │   ├── _index -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/_index
    │   ├── index -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index
    │   └── index.py -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/external/rules_pydeps+/pydeps/private/index/index.py
    └── py
        ├── __init__.py -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/external/rules_pydeps+/pydeps/private/py/__init__.py
        ├── python_module.py -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/external/rules_pydeps+/pydeps/private/py/python_module.py
        ├── python_source.py -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/external/rules_pydeps+/pydeps/private/py/python_source.py
        └── source_files.py -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/external/rules_pydeps+/pydeps/private/py/source_files.py

4 directories, 10 files
```

But when I inspect the execroot directories where the symlinks point, there is no `rules_pydeps+` directory:

```
❯ tree -L 1 /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/external/
/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/execroot/_main/external/
├── bazel_tools -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/bazel_tools
├── bazel_tools+xcode_configure_extension+local_config_xcode -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/bazel_tools+xcode_configure_extension+local_config_xcode
├── platforms -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/platforms
├── rules_cc+ -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_cc+
├── rules_cc++cc_configure_extension+local_config_cc -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_cc++cc_configure_extension+local_config_cc
├── rules_multitool++multitool+multitool -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_multitool++multitool+multitool
├── rules_multitool++multitool+multitool.macos_arm64 -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_multitool++multitool+multitool.macos_arm64
├── rules_mypy+ -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_mypy+
├── rules_python+ -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_python+
├── rules_python++pip+pypi -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_python++pip+pypi
├── rules_python++pip+pypi_312_click_py3_none_any_ae74fb96 -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_python++pip+pypi_312_click_py3_none_any_ae74fb96
├── rules_python++pip+rules_mypy_pip -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_python++pip+rules_mypy_pip
├── rules_python++pip+rules_mypy_pip_312_mypy_cp312_cp312_macosx_11_0_arm64_39bb21c6 -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_python++pip+rules_mypy_pip_312_mypy_cp312_cp312_macosx_11_0_arm64_39bb21c6
├── rules_python++pip+rules_mypy_pip_312_mypy_extensions_py3_none_any_4392f6c0 -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_python++pip+rules_mypy_pip_312_mypy_extensions_py3_none_any_4392f6c0
├── rules_python++pip+rules_mypy_pip_312_typing_extensions_py3_none_any_04e5ca03 -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_python++pip+rules_mypy_pip_312_typing_extensions_py3_none_any_04e5ca03
├── rules_python++python+python_3_12_aarch64-apple-darwin -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_python++python+python_3_12_aarch64-apple-darwin
└── rules_uv+ -> /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_uv+
```

It does looks like `/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_pydeps+`, so the process that constructs execroot seems to be missing a relevant link for some reason.
</details>

### Which category does this issue belong to?

Local Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


A small but perhaps not minimal repro here: https://github.com/theoremlp/bazel8_problem_repro

Changing .bazelversion to 7.4.1 and running `bazel build ...` succeeds. Running `bazel build ...` at version 8.0.0 fails with symlink problems as described above.

### Which operating system are you running Bazel on?


MacOS, Ubuntu

### What is the output of `bazel info release`?


release 8.0.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?


_No response_

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.

(see comment below, the initial bisect did not correctly identify the first bad commit)

### Have you found anything relevant by searching the web?

Unfortunately no.

### Any other information, logs, or outputs that you want to share?

_No response_

Ok, I've re-run backwards from 8.0.0 and the first commit where the build fails is 27487cf11f61eb23108f8df129dc1a00ee1a4370.

Verifying like this:
```
❯ USE_BAZEL_VERSION=e749da2f60025e2462f7cc7eea0457eb5734a51a bazelisk clean --expunge
INFO: Starting clean (this may take a while). Use --async if the clean takes more than several minutes.
❯ USE_BAZEL_VERSION=e749da2f60025e2462f7cc7eea0457eb5734a51a bazelisk build ...
Starting local Bazel server and connecting to it...
WARNING: Couldn't auto load rules or symbols, because no dependency on module/repository 'rules_android' found. This will result in a failure if there's a reference to those rules or symbols.
INFO: Analyzed 5 targets (107 packages loaded, 6240 targets configured).
INFO: Found 5 targets...
INFO: Elapsed time: 26.112s, Critical Path: 17.74s
INFO: 40 processes: 37 internal, 3 darwin-sandbox.
INFO: Build completed successfully, 40 total actions

❯ USE_BAZEL_VERSION=27487cf11f61eb23108f8df129dc1a00ee1a4370 bazelisk clean --expunge
Starting local Bazel server and connecting to it...
INFO: Starting clean (this may take a while). Use --async if the clean takes more than several minutes.
❯ USE_BAZEL_VERSION=27487cf11f61eb23108f8df129dc1a00ee1a4370 bazelisk build ...
Starting local Bazel server and connecting to it...
WARNING: Couldn't auto load rules or symbols, because no dependency on module/repository 'rules_android' found. This will result in a failure if there's a reference to those rules or symbols.
INFO: Analyzed 5 targets (107 packages loaded, 6240 targets configured).
ERROR: /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_pydeps++reqs+reqs/BUILD.bazel:7:15: Action external/rules_pydeps++reqs+reqs/pip_deps_index failed: (Exit 1): index failed: error executing Action command (from target @@rules_pydeps++reqs+reqs//:pip_deps_index) bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index ... (remaining 1 argument skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/1/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index", line 580, in <module>
    Main()
  File "/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/1/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index", line 504, in Main
    assert os.path.exists(main_filename), \
AssertionError: Cannot exec() '/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/1/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index.runfiles/_main/../rules_pydeps+/pydeps/private/index/index.py': file not found.
INFO: Elapsed time: 14.971s, Critical Path: 2.08s
INFO: 39 processes: 39 internal.
ERROR: Build did NOT complete successfully
```

I think when I tested this on master originally it was succeeding because of a cache. It still appears to be failing on master, though:
```
❯ USE_BAZEL_VERSION=b09335b86e03732d6da2fe40a7ac07af48480771 bazelisk clean --expunge
Starting local Bazel server (no_version) and connecting to it...
INFO: Starting clean (this may take a while). Use --async if the clean takes more than several minutes.
❯ USE_BAZEL_VERSION=b09335b86e03732d6da2fe40a7ac07af48480771 bazelisk build ...
Starting local Bazel server (no_version) and connecting to it...
WARNING: Couldn't auto load rules or symbols, because no dependency on module/repository 'rules_android' found. This will result in a failure if there's a reference to those rules or symbols.
INFO: Analyzed 5 targets (107 packages loaded, 6240 targets and 26 aspects configured).
ERROR: /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_pydeps++reqs+reqs/BUILD.bazel:7:15: Action external/rules_pydeps++reqs+reqs/pip_deps_index failed: (Exit 1): index failed: error executing Action command (from target @@rules_pydeps++reqs+reqs//:pip_deps_index) bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index ... (remaining 1 argument skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/1/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index", line 580, in <module>
    Main()
  File "/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/1/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index", line 504, in Main
    assert os.path.exists(main_filename), \
AssertionError: Cannot exec() '/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/1/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index.runfiles/_main/../rules_pydeps+/pydeps/private/index/index.py': file not found.
INFO: Elapsed time: 15.820s, Critical Path: 2.02s
INFO: 47 processes: 47 internal.
ERROR: Build did NOT complete successfully
```

On master, the most recent 'good' commit was eae9d4d019ca8d01bd3db5a8549b7f0e3505db61, and the subsequent 'bad' commit was 54237be06080d16a50f9c9ab0ad3a8f2651ca4d2.

```
❯ USE_BAZEL_VERSION=eae9d4d019ca8d01bd3db5a8549b7f0e3505db61 bazelisk clean --expunge
Starting local Bazel server (no_version) and connecting to it...
INFO: Starting clean (this may take a while). Use --async if the clean takes more than several minutes.
❯ USE_BAZEL_VERSION=eae9d4d019ca8d01bd3db5a8549b7f0e3505db61 bazelisk build ...
Starting local Bazel server (no_version) and connecting to it...
WARNING: Couldn't auto load rules or symbols, because no dependency on module/repository 'rules_android' found. This will result in a failure if there's a reference to those rules or symbols.
INFO: Analyzed 5 targets (107 packages loaded, 6240 targets and 26 aspects configured).
INFO: Found 5 targets...
INFO: Elapsed time: 25.012s, Critical Path: 14.97s
INFO: 48 processes: 45 internal, 3 darwin-sandbox.
INFO: Build completed successfully, 48 total actions
```

```
❯ USE_BAZEL_VERSION=54237be06080d16a50f9c9ab0ad3a8f2651ca4d2 bazelisk clean --expunge
Starting local Bazel server (no_version) and connecting to it...
INFO: Starting clean (this may take a while). Use --async if the clean takes more than several minutes.
❯ USE_BAZEL_VERSION=54237be06080d16a50f9c9ab0ad3a8f2651ca4d2 bazelisk build ...
Starting local Bazel server (no_version) and connecting to it...
WARNING: Couldn't auto load rules or symbols, because no dependency on module/repository 'rules_android' found. This will result in a failure if there's a reference to those rules or symbols.
INFO: Analyzed 5 targets (107 packages loaded, 6240 targets and 26 aspects configured).
ERROR: /private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/external/rules_pydeps++reqs+reqs/BUILD.bazel:7:15: Action external/rules_pydeps++reqs+reqs/pip_deps_index failed: (Exit 1): index failed: error executing Action command (from target @@rules_pydeps++reqs+reqs//:pip_deps_index) bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index ... (remaining 1 argument skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/1/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index", line 580, in <module>
    Main()
  File "/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/1/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index", line 504, in Main
    assert os.path.exists(main_filename), \
AssertionError: Cannot exec() '/private/var/tmp/_bazel_melliot/ec7f0e67bfc3fe3a9bd546beff97b5c5/sandbox/darwin-sandbox/1/execroot/_main/bazel-out/darwin_arm64-opt-exec-ST-c6280171b4e4/bin/external/rules_pydeps+/pydeps/private/index/index.runfiles/_main/../rules_pydeps+/pydeps/private/index/index.py': file not found.
INFO: Elapsed time: 12.961s, Critical Path: 2.29s
INFO: 47 processes: 47 internal.
ERROR: Build did NOT complete successfully
```

--- a/src/main/java/com/google/devtools/build/lib/skyframe/BuildDriverFunction.java
+++ b/src/main/java/com/google/devtools/build/lib/skyframe/BuildDriverFunction.java
@@ -324,6 +324,7 @@ public class BuildDriverFunction implements SkyFunction {
       ImmutableSet.Builder<Artifact> artifactsToBuild = ImmutableSet.builder();
       List<SkyKey> aspectCompletionKeys = new ArrayList<>();
 
+      // Do not trigger Skyframe restarts in this loop (see comments below).
       for (Map.Entry<AspectKey, AspectValue> entry :
           ((TopLevelAspectsValue) topLevelSkyValue).getTopLevelAspectsMap().entrySet()) {
         AspectKey aspectKey = entry.getKey();
@@ -339,13 +340,20 @@ public class BuildDriverFunction implements SkyFunction {
         NestedSet<Package> transitivePackagesForSymlinkPlanting =
             aspectValue.getTransitivePackages();
         if (transitivePackagesForSymlinkPlanting != null) {
-          postEventIfNecessary(
-              postedEventsTypes,
-              env,
-              TopLevelTargetReadyForSymlinkPlanting.create(transitivePackagesForSymlinkPlanting));
+          // This event should be sent out exactly once per aspect in this BuildDriverKey, even with
+          // resets. We achieve this by marking the event type as sent only after sending the event
+          // for all aspects, but must avoid triggering Skyframe restarts while doing so.
+          if (!postedEventsTypes.contains(
+              TopLevelStatusEvents.Type.TOP_LEVEL_TARGET_READY_FOR_SYMLINK_PLANTING)) {
+            env.getListener()
+                .post(
+                    TopLevelTargetReadyForSymlinkPlanting.create(
+                        transitivePackagesForSymlinkPlanting));
+          }
         }
         aspectCompletionKeys.add(AspectCompletionKey.create(aspectKey, topLevelArtifactContext));
       }
+      postedEventsTypes.add(TopLevelStatusEvents.Type.TOP_LEVEL_TARGET_READY_FOR_SYMLINK_PLANTING);
 
       if (!additionalPostAnalysisDepsRequestedAndAvailable.request(env, actionLookupKey)) {
         return null;

